common/web/utils/src/timeoutPromise.ts

```diff
@@ -0,0 +1,59 @@
+import ManagedPromise from "./managedPromise.js";
+
+/**
+ * This class represents a cancelable timeout, wrapped in Promise form.
+ *
+ * It will resolve to `true` when the timer completes unless `resolve` or
+ * `reject` is called earlier.  Call `.resolve(false)` for early cancellation
+ * or `.resolve(true)` to cancel the timer while resolving the Promise early.
+ */
+export default class TimeoutPromise extends ManagedPromise<Boolean> {
+  private timerHandle: number | NodeJS.Timeout;
+  constructor(timeoutInMillis: number) {
+    // Helps marshal the internal timer handle to its member field despite being
+    // initialized in a closure passed to `super`, which cannot access `this`.
+    const timerHandleCapture = {
+      val: null as number | NodeJS.Timeout
+    };
+
+    super((resolve) => {
+      const timerId = setTimeout(() => {
+        if(!this.isResolved) {
+          resolve(true)
+        }
+      }, timeoutInMillis);
+
+      // Forwards the timer handle outside of the closure.
+      timerHandleCapture.val = timerId;
+    });
+
+    // "Lands" the timer handle in its final destination.
+    this.timerHandle = timerHandleCapture.val;
+
```

Comment:
Shouldn't the following work? 
```suggestion
    let timerHandleCapture: null as number | NodeJS.Timeout = null;

    super((resolve) => {
      const timerId = setTimeout(() => {
        if(!this.isResolved) {
          resolve(true)
        }
      }, timeoutInMillis);

      // Forwards the timer handle outside of the closure.
      timerHandleCapture = timerId;
    });

    // "Lands" the timer handle in its final destination.
    this.timerHandle = timerHandleCapture;

```

Reply:
That does seem to work, yes.

In some languages that extra layer would be needed, so I'm always a little skittish about leaning into the pattern.